shanoir-ng-datasets/src/main/java/org/shanoir/ng/dicom/web/StudyInstanceUIDHandler.java

```diff
@@ -0,0 +1,174 @@
+					JsonNode retrieveURLNode = root.get(fieldName);
+					ArrayNode retrieveURLArray = (ArrayNode) retrieveURLNode.path("Value");
+					for (int i = 0; i < retrieveURLArray.size(); i++) {
+						JsonNode arrayElement = retrieveURLArray.get(i);
+						String retrieveURL = arrayElement.asText();
+						if (studyLevel) { // study level
+							retrieveURL = retrieveURL.replaceFirst("/studies/(.*)", "/studies/" + examinationId);
+							retrieveURLArray.remove(i);
+							retrieveURLArray.add(retrieveURL);
+						} else { // serie level
+							retrieveURL = retrieveURL.replaceFirst("/studies/(.*)/series/", "/studies/" + examinationId + "/series/");
+							retrieveURLArray.remove(i);
+							retrieveURLArray.add(retrieveURL);
+						}
+					}
+				}
+			}
+		} else if (root.isArray()) {
+			ArrayNode arrayNode = (ArrayNode) root;
+			for (int i = 0; i < arrayNode.size(); i++) {
+				JsonNode arrayElement = arrayNode.get(i);
+				replaceStudyInstanceUIDsWithExaminationIds(arrayElement, examinationId, studyLevel);
+			}
+		}
+	}
+
+	/**
+	 * This method returns the corresponding StudyInstanceUID, that is generated during the import in Shanoir
+	 * with the pseudonymization module and present in the PACS, either from a local cache to accelerate the
+	 * request response time or from the database, in table dataset_file.
+	 * 
+	 * @param examinationId
+	 * @return
+	 */
+	public String findStudyInstanceUIDFromCacheOrDatabase(Long examinationId) {
+		String studyInstanceUID = examinationIdToStudyInstanceUIDCache.get(examinationId);
+		if (studyInstanceUID == null) {
+			Examination examination = examinationService.findById(examinationId);
+			if (examination != null) {
+				studyInstanceUID = findStudyInstanceUID(examination);
+				if (studyInstanceUID != null) {
+					examinationIdToStudyInstanceUIDCache.put(examination.getId(), studyInstanceUID);
+					LOG.info("DICOMWeb cache adding: " + examinationId + ", " + studyInstanceUID);
+					LOG.info("DICOMWeb cache, size: " + examinationIdToStudyInstanceUIDCache.size());
+				}
+			}
+		}
+		return studyInstanceUID;
+	}
+	
+	/**
+	 * This method walks down the information model in Shanoir to read the StudyInstanceUID
+	 * from the table dataset_file.path, that contains the WADO link.
+	 * 
+	 * @param examination
+	 * @return
+	 */
+	private String findStudyInstanceUID(Examination examination) {
+		List<DatasetAcquisition> acquisitions = examination.getDatasetAcquisitions();
+		if (!acquisitions.isEmpty()) {
+			DatasetAcquisition acquisition = acquisitions.get(0);
+			List<Dataset> datasets = acquisition.getDatasets();
+			if (!datasets.isEmpty()) {
+				Dataset dataset = datasets.get(0);
+				List<DatasetExpression> expressions = dataset.getDatasetExpressions();
+				if (!expressions.isEmpty()) {
+					for (DatasetExpression expression : expressions) {
+						// only DICOM is of interest here
+						if (expression.getDatasetExpressionFormat().equals(DatasetExpressionFormat.DICOM)) {
```

Comment:
If we have an examination with an EEG acquisition as first acquisition, we won't retrieve anything. I think we need to iterated over acquisitions / datasets until we eventually find one containing DICOMs.

Maybe add a check on datasetFile.isPacs() to be sure that we are retrieving a dicom path from the PACS and not from the file system ? (and iterate over files too in this way)

Reply:
We actually have many weird things in the DB (we have an issue for this): 
select count(*) from dataset_file df JOIN dataset_expression de ON df.dataset_expression_id = de.id WHERE de.dataset_expression_format = 6 and df.pacs is false;
+----------+
| count(*) |
+----------+
|    20414 |
+----------+

So the check on "pacs == true" should totally be done :)

Reply:
- EEG acquisitions are not DICOM and can not be displayed by OHIF-viewer, normally
- great, THX!, I have added the isPacs() check to be protected